### ISSUE TITLE
HTTP2/ssl errors

### ISSUE THREAD

**Issue by cjhawley** [facebook/proxygen#158]
Hi,

I am trying to call the sample EchoServer with HTTP2 using curl

`curl -vkI --http2 https://localhost:11002`

and am receiving the following error:

```
* Rebuilt URL to: https://localhost:11002/                                                            │ -T, --upload-file FILE  Transfer FILE to destination
*   Trying ::1...                                                                                     │     --url URL       URL to work with
* TCP_NODELAY set                                                                                     │ -B, --use-ascii     Use ASCII/text transfer
* Connected to localhost (::1) port 11002 (#0)                                                        │ -u, --user USER[:PASSWORD]  Server user and password
* ALPN, offering h2                                                                                   │     --tlsuser USER  TLS username
* ALPN, offering http/1.1                                                                             │     --tlspassword STRING  TLS password
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH                        │     --tlsauthtype STRING  TLS authentication type (default: SRP)
* successfully set certificate verify locations:                                                      │     --unix-socket PATH    Connect through this Unix domain socket
*   CAfile: /usr/local/etc/openssl/cert.pem                                                           │     --abstract-unix-socket PATH Connect to an abstract Unix domain socket
  CApath: /usr/local/etc/openssl/certs                                                                │ -A, --user-agent STRING  Send User-Agent STRING to server (H)
* TLSv1.2 (OUT), TLS header, Certificate Status (22):                                                 │ -v, --verbose       Make the operation more talkative
* TLSv1.2 (OUT), TLS handshake, Client hello (1):                                                     │ -V, --version       Show version number and quit
* error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol                                 │ -w, --write-out FORMAT  Use output FORMAT after completion
* Closing connection 0                                                                                │     --xattr         Store metadata in extended file attributes
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

The logs on the server are:

```
E0328 17:39:47.926362     9 HTTP2Codec.cpp:1312] Connection error with ingress=
00000000  16 03 01 02 00 01 00 01  fc 03 03 7c c1 df 0c db  |...........|....|
00000010  27 9b a4 69 55 59 7e a1  a5 ba a2 ee 2e 58 8d 61  |'..iUY~......X.a|
00000020  90 19 f2 f6 94 9c 0d a3  fb e6 51 00 00 a0 c0 30  |..........Q....0|
00000030  c0 2c c0 28 c0 24 c0 14  c0 0a 00 a5 00 a3 00 a1  |.,.(.$..........|
00000040  00 9f 00 6b 00 6a 00 69  00 68 00 39 00 38 00 37  |...k.j.i.h.9.8.7|
00000050  00 36 00 88 00 87 00 86  00 85 c0 32 c0 2e c0 2a  |.6.........2...*|
00000060  c0 26 c0 0f c0 05 00 9d  00 3d 00 35 00 84 c0 2f  |.&.......=.5.../|
00000070  c0 2b c0 27 c0 23 c0 13  c0 09 00 a4 00 a2 00 a0  |.+.'.#..........|
00000080  00 9e 00 67 00 40 00 3f  00 3e 00 33 00 32 00 31  |...g.@.?.>.3.2.1|
00000090  00 30 00 9a 00 99 00 98  00 97 00 45 00 44 00 43  |.0.........E.D.C|
000000a0  00 42 c0 31 c0 2d c0 29  c0 25 c0 0e c0 04 00 9c  |.B.1.-.).%......|
000000b0  00 3c 00 2f 00 96 00 41  00 07 c0 12 c0 08 00 16  |.<./...A........|
000000c0  00 13 00 10 00 0d c0 0d  c0 03 00 0a 00 ff 01 00  |................|
000000d0  01 33 00 00 00 0e 00 0c  00 00 09 6c 6f 63 61 6c  |.3.........local|
000000e0  68 6f 73 74 00 0b 00 04  03 00 01 02 00 0a 00 1c  |host............|
000000f0  00 1a 00 17 00 19 00 1c  00 1b 00 18 00 1a 00 16  |................|
00000100  00 0e 00 0d 00 0b 00 0c  00 09 00 0a 00 0d 00 20  |............... |
00000110  00 1e 06 01 06 02 06 03  05 01 05 02 05 03 04 01  |................|
00000120  04 02 04 03 03 01 03 02  03 03 02 01 02 02 02 03  |................|
00000130  00 0f 00 01 01 33 74 00  00 00 10 00 0e 00 0c 02  |.....3t.........|
00000140  68 32 08 68 74 74 70 2f  31 2e 31 00 15 00 b6 00  |h2.http/1.1.....|
00000150  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000160  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000170  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000180  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000190  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001a0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000001f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000200  00 00 00 00 00                                    |.....           |
```

setting up proxygen via Docker, on latest master.

Any ideas? I've been trying to work on it for a bit, but haven't had any leads.

Thanks!
Chris

**Comment by lnicco** [comment]
@cjhawley the sample server is not doing SSL, so that port expects an upgrade on cleartext.

With Curl you can get that by using the `--http2-prior-knowledge` option 

```
$ curl -I --http2 --http2-prior-knowledge http://localhost:11002
HTTP/2 200
request-number: 1
content-length: 0
date: Wed, 29 Mar 2017 20:41:38 GMT
```

I am using latest curl: curl-7.53.1

**Comment by cjhawley** [comment]
Thanks!

Also, I should have looked more closely to the issues -- this enables ssl on the sample server: https://github.com/facebook/proxygen/issues/65

**Comment by abogani** [comment]
@lnicco Is there an example on how add support for the upgrade on cleartext?

Thanks!

**Comment by afrind** [comment]
See proxygen/httpclient/samples/curl with the -h2c option

**Comment by abogani** [comment]
@afrind Unfortunately Proxygen (as server) doesn't upgrade to h2. I have tried with Chrome and Firefox browsers, curl and samples/curl (https://groups.google.com/d/msg/facebook-proxygen/kI89bZSe3sE/mJ3oP3YaAQAJ). I don't figured out yet what I have done wrong.

**Comment by afrind** [comment]
Proxygen’s sample echo server is configured to perform plaintext upgrade when requested.

EchoServer.cpp:78:  options.h2cEnabled = true;

Run curl -http2 http://localhost:11000 to see it in action.  Maybe the push sample doesn’t enable plaintext upgrade?

**Comment by abogani** [comment]
@afrind I'm not be able to make protocol upgrade works with Proxygen: none of the sample does the tricks for me with and without SSL, h2cEnabled and allowInsecureConnectionsOnSecureServer. So no HTTP2 here.

`curl -kv --http2 https://localhost:11002` says:
[...]
* ALPN, offering h2
* ALPN, offering http/1.1
[...]
* ALPN, server did not agree to a protocol <---- This is odd!
[...]
> GET / HTTP/1.1 <------ Should this be HTTP/2.0?
[...]
< HTTP/1.1 200 OK

Proxygen says:
[...]
I0404 08:20:50.081094 29350 SSLAcceptorHandshakeHelper.cpp:78] Client did not select a next protocol
[...]
I0404 08:20:50.082099 29350 HTTPMessage.cpp:602] , chunked: 0, upgraded: 0, Fields 
                                                                                                            ^^^^^^^^ 

Should be first message an upgrade one?

Use Chrome's Developer Tools confirm curl: Proxygen (my installation at least) doesn't make protocol upgrade (their report http/1.1 instead of h2). 

Do you have any suggestions?

Thanks a lot!

**Comment by afrind** [comment]
Port 11000 is http/1.1, cleartext – this is the one you can do upgrade on, run curl -http2 http://localhost:11000
Port 11002 is normally cleartext http/2 – you can only curl this with --http2-prior-knowledge

If you are using TLS (looks like you are) then you have to negotiate H2 with an ALPN token, and you cannot upgrade.  You probably need to call setNextProtocols on your SSLContextConfig.

As for seeing 1.1, well you are doing 1.1 so that is expected.   Per the spec “HTTP/2 does not define a way to carry the version identifier that is included in the HTTP/1.1 request line.”
